src/lib/kdf/xmd/xmd.cpp

```diff
@@ -0,0 +1,81 @@
+/*
+* (C) 2019,2020,2021,2024 Jack Lloyd
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/internal/xmd.h>
+
+#include <botan/exceptn.h>
+#include <botan/hash.h>
+#include <botan/internal/fmt.h>
+#include <vector>
+
+namespace Botan {
+
+void expand_message_xmd(std::string_view hash_fn,
+                        std::span<uint8_t> output,
+                        std::span<const uint8_t> input,
+                        std::span<const uint8_t> domain_sep) {
+   if(domain_sep.size() > 0xFF) {
+      // RFC 9380 has a specification for handling this
+      throw Not_Implemented("XMD does not currently implement oversize DST handling");
+   }
+
+   const uint8_t domain_sep_len = static_cast<uint8_t>(domain_sep.size());
+
+   auto hash = HashFunction::create_or_throw(hash_fn);
+   const size_t block_size = hash->hash_block_size();
+   if(block_size == 0) {
+      throw Invalid_Argument(fmt("expand_message_xmd cannot be used with {}", hash_fn));
+   }
+
+   const size_t hash_output_size = hash->output_length();
+   if(output.size() > 255 * hash_output_size || output.size() > 0xFFFF) {
+      throw Invalid_Argument("expand_message_xmd requested output length too long");
+   }
+
+   // Compute b_0 = H(msg_prime) = H(Z_pad || msg || l_i_b_str || 0x00 || DST_prime)
+
+   hash->update(std::vector<uint8_t>(block_size));
+   hash->update(input);
+   hash->update_be(static_cast<uint16_t>(output.size()));
+   hash->update(0x00);
+   hash->update(domain_sep);
+   hash->update(domain_sep_len);
+
+   const secure_vector<uint8_t> b_0 = hash->final();
+
+   // Compute b_1 = H(b_0 || 0x01 || DST_prime)
+
+   hash->update(b_0);
+   hash->update(0x01);
+   hash->update(domain_sep);
+   hash->update(domain_sep_len);
+
+   secure_vector<uint8_t> b_i = hash->final();
+
+   uint8_t cnt = 2;
+   for(;;) {
+      const size_t produced = std::min(output.size(), hash_output_size);
+
+      copy_mem(&output[0], b_i.data(), produced);
```

Comment:
Maybe:

```suggestion
      copy_mem(std::span<uint8_t>(output).first(produced), std::span<const uint8_t>(b_i).first(produced));
```

... I've used this idiom in some places, but I'm not super happy with it. It's quite chatty, and IIUC `.first()` doesn't even range-check the `span`, so technically it isn't even saver than the bare pointer stuff.

Reply:
Lot of noise with not a lot of benefit imo. Probably worth investigating options for some span-centric operations like this for the general case.